### PR TITLE
8332086: Remove the usage of ServiceLoader in j.u.r.RandomGeneratorFactory

### PR DESCRIPTION
**Pull request by rgiulietti**
All random number generator algorithms are implemented in module `java.base`. The usage of `ServiceLoader` in `j.u.r.RandomGeneratorFactory` is no longer needed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8332477](https://bugs.openjdk.org/browse/JDK-8332477) to be approved

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332476: j.u.r.RandomGeneratorFactor.create(long|byte[]) should throw rather than silently fallback to no-arg create()`

### Issues
 * [JDK-8332086](https://bugs.openjdk.org/browse/JDK-8332086): Remove the usage of ServiceLoader in j.u.r.RandomGeneratorFactory (**Bug** - P4)
 * [JDK-8332476](https://bugs.openjdk.org/browse/JDK-8332476): j.u.r.RandomGeneratorFactor.create(long|byte[]) should throw rather than silently fallback to no-arg create() (**Bug** - P4)
 * [JDK-8332131](https://bugs.openjdk.org/browse/JDK-8332131): Remove the usage of ServiceLoader in j.u.r.RandomGeneratorFactory (**CSR**) (Withdrawn)
 * [JDK-8332477](https://bugs.openjdk.org/browse/JDK-8332477): j.u.r.RandomGeneratorFactor.create(long|byte[]) should throw rather than silently fallback to no-arg create() (**CSR**)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19212/head:pull/19212` \
`$ git checkout pull/19212`

Update a local copy of the PR: \
`$ git checkout pull/19212` \
`$ git pull https://git.openjdk.org/jdk.git pull/19212/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19212`

View PR using the GUI difftool: \
`$ git pr show -t 19212`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19212.diff">https://git.openjdk.org/jdk/pull/19212.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19212#issuecomment-2107009978)

### CODE REVIEW

**Comment by rgiulietti** [comment]
This is a followup of [18932](https://github.com/openjdk/jdk/pull/18932) in which all random number generator algorithms have been moved to module `java.base` and module `jdk.random` has been removed.

Reliance on `ServiceLoader` in `j.u.r.RandomGeneratorFactory` is no longer needed and has been replaced by a (lazily populated) map from algorithm names to classes.

Moreover, methods `RandomGeneratorFactory.create(long)` and `create(byte[])` now throw an `UnsupportedOperationException` rather than silently falling back to the no-arg `create()` in case the underlying algorithm does not support a `long` resp. `byte[]` seed.

Tests in tier1-tier3 pass after adaptations on the existing `RandomTestCoverage`.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 1]
```diff
@@ -108,78 +102,93 @@
...
-        if (properties == null) {
-            synchronized (provider) {
-                if (properties == null) {
-                    properties = provider.type().getDeclaredAnnotation(RandomGeneratorProperties.class);
-                    Objects.requireNonNull(properties, provider.type() + " missing annotation");
+        if (properties == null) {  // volatile load
+            synchronized (rgClass) {
+                if (properties == null) {  // double-checking idiom
+                    RandomGeneratorProperties props = rgClass.getDeclaredAnnotation(RandomGeneratorProperties.class);
+                    Objects.requireNonNull(props, rgClass + " missing annotation");
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: Hello Raffaello, with the `RandomGenerator` implementations now all residing within the java.base module, I think an additional advantage of that is that, it will now allow us to remove this internal `RandomGeneratorProperties` annotation and thus this reflection code.

I think one way to do that would be something like this within this `RandomGeneratorFactory` class itself: 

```java
private record RandomGenEntry(Class<?> randomGenClass, int i, int j,
                              int k, int equiDistribution, boolean stochastic,
                              boolean hardware) {

}

private static final Map<String, RandomGenEntry> FACTORY_MAP = ... // construct the map

```
where the `FACTORY_MAP` will be keyed to the alogrithm and the value will be a record which holds these additional details about the `RandomGenerator`. 
This current PR is about getting rid of ServiceLoader usage. So if you want to remove the usage of this annotation and reflection is a separate PR that's fine with me. Furthermore, although I don't see the necessity of an annotation for what we are doing here, if you think that the removal of the annotation and reflection isn't worth doing, that is OK too.

Comment (jaikiran) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: Thinking a bit more, I think we can even get rid of the reflection used in `create()` methods implementation, if we wanted to, by doing something like this:

```java
private record RandomGenEntry(Class<? extends RandomGenerator> randomGenClass, int i, int j,
                                  int k, int equiDistribution, boolean stochastic,
                                  boolean hardware) {

        RandomGenerator create() {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random" -> new Random();
                case "L128X1024MixRandom" -> new L128X1024MixRandom();
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus();
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }

        RandomGenerator create(long seed) {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random", "SplittableRandom", "SecureRandom" -> {
                    throw new UnsupportedOperationException("cannot construct with a long seed");
                }
                case "L128X1024MixRandom" -> new L128X1024MixRandom(seed);
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus(seed);
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }

        RandomGenerator create(byte[] seed) {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random", "SplittableRandom", "SecureRandom" -> {
                    throw new UnsupportedOperationException("cannot construct with a byte[] seed");
                }
                case "L128X1024MixRandom" -> new L128X1024MixRandom(seed);
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus(seed);
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }
    }

    
    private static final Map<String, RandomGenEntry> FACTORY_MAP = ... // construct the map

```

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: @jaikiran I agree that we can simplify even more. I just wanted to change as little as possible in this PR to facilitate reviews.
Shall I push your proposed changes in this PR or is a followup PR preferable?

Comment (jaikiran) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: A followup PR is fine. I think once this PR which addresses the API aspects (like removal of ServiceLoader and then updates to the create() method javadoc) is integrated, then the subsequent PR can just be all internal implementation changes like the proposed removal of reflection.

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: Fine with me.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 2]
```diff
@@ -108,78 +102,93 @@
...
+     * @return RandomGeneratorProperties annotation for the specified random generator.
      */
      private RandomGeneratorProperties getProperties() {
-        if (properties == null) {
-            synchronized (provider) {
-                if (properties == null) {
-                    properties = provider.type().getDeclaredAnnotation(RandomGeneratorProperties.class);
-                    Objects.requireNonNull(properties, provider.type() + " missing annotation");
+        if (properties == null) {  // volatile load
+            synchronized (rgClass) {
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:187]: The synchronization on `rgClass` to intialize an instance field `properties` appears odd here. I think this should be synchronized on `this`.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 3]
```diff
@@ -273,70 +287,54 @@ static <T extends RandomGenerator> T of(String name, Class<T> category)
...
-                                Class<?> argType = parameterTypes[0];
-
-                                if (argType == long.class) {
-                                    tmpCtorLong = ctorSpecific;
-                                } else if (argType == byte[].class) {
-                                    tmpCtorBytes = ctorSpecific;
-                                }
+    private void ensureConstructors() {
+        if (ctor == null) {  // volatile load
+            synchronized (rgClass) {
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:299]: Here too I think we should synchronize on `this` - we would want to allow multiple different instances of a `RandomGeneratorFactory` for the same `RandomGenerator` class type to be able to concurrently instantiate their individual instance fields (like the `ctor`(s) and `properties`).

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:299]: Then I would even remove the double-checking idiom, the `volatile` on `ctor` and `properties`, and declare methods `getProperties()` and `ensureConstructors()` as `synchronized`.
I'm not sure that the double-checking optimization brings much value on contemporary JVMs.

But I feel that the followup PR discussed before wouldn't need `synchronized` at all.

WDYT?

Comment (jaikiran) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:299]: > Then I would even remove the double-checking idiom, the volatile on ctor and properties, and declare methods getProperties() and ensureConstructors() as synchronized.
> I'm not sure that the double-checking optimization brings much value on contemporary JVMs.

Making the methods synchronized would bring in a penalty that there will always be a monitor entry at every call site, even after the `properites` and `ctor`(s) are initialized. Ideally, we should just do all of this intialization in the constructor of the `RandomGeneratorFactory`, the one which takes the `Class<>` type of the `RandomGenerator`. We can then make the `properties` and the `ctor`(s) all `final` and not have to worry about any synchronization or volatile semantics. You would of course have to rework the ensureConstructors to not throw an exception at that time.

> But I feel that the followup PR discussed before wouldn't need synchronized at all.

Correct. The more I think about it, I think cleaning up all this in this PR itself might make both reviewing and the implementation a bit more simpler. What's your thoughts?

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:299]: OK, will do all the work in this PR.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 1 above)
Comment (dfuch) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: Not sure if that's an issue - but if you wanted/needed to delay the loading of those random generator classes that will not be used until something actually wants to use them, you could consider using a `Supplier<Class<? extends RandomGenerator>>` or a `Supplier<RandomGenEntry>` for the `FACTORY_MAP` values?

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: @dfuch You mean not loading the whole batch but only individual classes, as need arises?

Comment (dfuch) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: yes - I do not know if loading all the classes in one batch could be an issue at startup (I'd expect Random to be loaded early) - but if it proves to be, then using a Supplier to load them on demand might do the trick. If creating Random or SecureRandom does not trigger this code - and if those classes are not loaded at startup - then maybe that's a non issue and you can just ignore my comment.

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:190]: Loading of `Random`, `SecureRandom`, `SplittableRandom`, or `ThreadLocalRandom` does not trigger the loading of `RandomGeneratorFactory`.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 4]
```diff
@@ -107,137 +114,263 @@
...
+                            64, 0, 0, 1,
+                            0)
+            );
+        }
+
+        private static SimpleImmutableEntry<String, RandomGeneratorProperties>
+        entry(Class<? extends RandomGenerator> rgClass, String name, String group,
+                int i, int j, int k, int equidistribution,
+                int flags) {
+            return new SimpleImmutableEntry<>(name,
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:202]: This should use [`Map​.Entry​<…>`] as the return type, which in turn allows using [`Map​::entry​(K, V)`] as the implementation, as neither the `name` key, nor the value is ever `null`:
```suggestion
        private static Map.Entry<String, RandomGeneratorProperties>
        entry(Class<? extends RandomGenerator> rgClass, String name, String group,
                int i, int j, int k, int equidistribution,
                int flags) {
            return Map.entry(name,
```

[`Map​.Entry​<…>`]: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/Map.Entry.html
[`Map​::entry​(K, V)`]: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/util/Map.html#entry(K,V)

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java]: @ExE-Boss Thanks for the suggestion.

**Comment by jaikiran** [comment]
Thank you for these updates Raffaello. I think the implementation is looking much more simpler and cleaner now. I've a few documentation related review comments which I've added inline. I haven't reviewed the test change yet.

File: src/java.base/share/classes/java/util/random/package-info.java [hunk 5]
```diff
@@ -81,8 +81,8 @@
  * <blockquote>{@code import java.util.random.*;}</blockquote>
  *
  * Then one can choose a specific implementation by giving the name of a generator
- * algorithm to the static method {@link RandomGenerator#of}, in which case the
- * no-arguments constructor for that implementation is used:
+ * algorithm to the static method {@link RandomGenerator#of}, in which case no
+ * seed is specified by the caller:
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/package-info.java:85]: Perhaps reword this to:


> ... in which case a {@code RandomGenerator} is constructed without any seed value:

Comment (jaikiran) [reply src/java.base/share/classes/java/util/random/package-info.java:85]: An already existing issue in the specification in this file is that, a few lines below, we note that:

> <p> There are three groups of random number generator algorithm provided in Java: the Legacy group, the LXM group, and the Xoroshiro/Xoshiro group.

The "three" groups is misleading I think, since both later in the table as well as the `group()` method implementation on `RandomGeneratorFactory`, we return four distinct values "Legacy", "LXM", "Xoroshiro" and "Xoshiro". Should we reword this part of the documentation to remove the mention of "three"?

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 6]
```diff
@@ -547,66 +603,59 @@ public boolean isStreamable() {
...
     /**
-     * Create an instance of {@link RandomGenerator} based on
-     * <a href="package-summary.html#algorithms">algorithm</a> chosen
-     * providing a starting long seed. If long seed is not supported by an
-     * algorithm then the no argument form of create is used.
+     * Create an instance of {@link RandomGenerator} based on the
+     * <a href="package-summary.html#algorithms">algorithm</a> chosen,
+     * and providing a starting long seed.
+     * If a long seed is not supported by the algorithm,
+     * an {@link UnsupportedOperationException} is thrown.
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:626]: Perhaps reword this and the other `create(byte[])` method to something like:

```
* Create an instance of {@link RandomGenerator} based on the
* <a href="package-summary.html#algorithms">algorithm</a> chosen,
* and the provided {@code seed}.
* If the {@code RandomGenerator} doesn't support instantiation through
* a {@code seed} of type {@code long} then this method throws an
* an {@link UnsupportedOperationException}.
```

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java [hunk 7]
```diff
@@ -107,137 +113,263 @@
...
+        }
+
+        private static Map.Entry<String, RandomGeneratorProperties> entry(
+                Class<? extends RandomGenerator> rgClass, String name, String group,
+                int i, int j, int k, int equidistribution,
+                int flags) {
+            return Map.entry(name,
+                    new RandomGeneratorProperties(rgClass, name, group,
+                            i, j, k, equidistribution,
+                            flags | (rgClass.isAnnotationPresent(Deprecated.class) ? DEPRECATED : 0)));
```
Comment (jaikiran) [inline src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:204]: Hello Raffaello, this is the final remaining reflection usage and even this I think isn't required now that all the random generator implementations reside within java.base as an implementation detail.

I think we should just skip this annotation check here and set `DEPRECATED` bit on the `flags` to `0` for all implementations. When/if we do deprecate any of the random generators, we can just come here and switch that bit to on for the specific random generator when instantiating this `RandomGeneratorProperties` record. I had a brief look at the code and the documentation in `package-info.java` of `java/util/random` and we don't mention that we rely on the `@Deprecated`  annotation to determine whether an algorithm is deprecated. I think that's a good thing.

File: test/jdk/java/util/Random/RandomTestCoverage.java [hunk 8]
```diff
@@ -171,8 +171,25 @@ static void coverFactory(RandomGeneratorFactory factory) {
...
 
         coverRandomGenerator(factory.create());
-        coverRandomGenerator(factory.create(12345L));
-        coverRandomGenerator(factory.create(new byte[] {1, 2, 3, 4, 5, 6, 7, 8}));
+        RandomGenerator rng;
+
+        rng = null;
+        try {
+            rng = factory.create(12345L);
+        } catch (UnsupportedOperationException ignore) {
```
Comment (jaikiran) [inline test/jdk/java/util/Random/RandomTestCoverage.java:179]: I think now that we know for sure which algorithm instances don't support which type of seed value and since throwing `UnsupportedOperationException` is now a specification of the `create(...)` methods, we should probably do something like this:

```diff
diff --git a/test/jdk/java/util/Random/RandomTestCoverage.java b/test/jdk/java/util/Random/RandomTestCoverage.java
index be12d188198..6e5c36e13c3 100644
--- a/test/jdk/java/util/Random/RandomTestCoverage.java
+++ b/test/jdk/java/util/Random/RandomTestCoverage.java
@@ -171,8 +171,37 @@ static void coverFactory(RandomGeneratorFactory factory) {
         boolean isSplittable = factory.isSplittable();
 
         coverRandomGenerator(factory.create());
-        coverRandomGenerator(factory.create(12345L));
-        coverRandomGenerator(factory.create(new byte[] {1, 2, 3, 4, 5, 6, 7, 8}));
+
+        String algo = factory.name();
+        // test create(long)
+        switch (algo) {
+            // SecureRandom doesn't have long constructors so we expect
+            // UnsupportedOperationException
+            case "SecureRandom" -> {
+                try {
+                    factory.create(12345L);
+                    throw new AssertionError("RandomGeneratorFactory.create(long) was expected" +
+                            "to throw UnsupportedOperationException for " + algo + " but didn't");
+                } catch (UnsupportedOperationException uoe) {
+                    // expected
+                }
+            }
+            default -> coverRandomGenerator(factory.create(12345L));
+        }
+        // test create(byte[])
+        switch (algo) {
+            // these don't have byte[] constructors so we expect UnsupportedOperationException
+            case "Random", "SplittableRandom" -> {
+                try {
+                    factory.create(new byte[] {1, 2, 3, 4, 5, 6, 7, 8});
+                    throw new AssertionError("RandomGeneratorFactory.create(byte[]) was expected" +
+                            "to throw UnsupportedOperationException for " + algo + " but didn't");
+                } catch (UnsupportedOperationException uoe) {
+                    // expected
+                }
+            }
+            default -> coverRandomGenerator(factory.create(new byte[] {1, 2, 3, 4, 5, 6, 7, 8}));
+        }
     }


```

File: test/jdk/java/util/Random/RandomTestCoverage.java [hunk 9]
```diff
@@ -171,8 +171,25 @@ static void coverFactory(RandomGeneratorFactory factory) {
...
+        try {
+            rng = factory.create(new byte[] {1, 2, 3, 4, 5, 6, 7, 8});
+        } catch (UnsupportedOperationException ignore) {
+        }
+        if (rng != null) {
+            coverRandomGenerator(rng);
+        }
     }
 
     static void coverDefaults() {
```
Comment (jaikiran) [inline test/jdk/java/util/Random/RandomTestCoverage.java:195]: This test method appears to test the calls to `getDefault()` methods on `RandomGeneratorFactory` and `RandomGenerator` classes, but it isn't being called in the test. We should call this method from `main()` to have test coverage for those methods.

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 7 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java:204]: Yes, I thought about this the other day but decided for a bit more conservative approach, relying on the annotation.

But I agree that, since the meta-information now resides in `RandomGeneratorProperties`, we might "migrate" the deprecation status here as well.

File: test/jdk/java/util/Random/RandomTestCoverage.java (hunk 8 above)
Comment (rgiulietti) [reply test/jdk/java/util/Random/RandomTestCoverage.java:179]: So you want to be very specific. OK.

File: test/jdk/java/util/Random/RandomTestCoverage.java (hunk 9 above)
Comment (rgiulietti) [reply test/jdk/java/util/Random/RandomTestCoverage.java:195]: Right

File: src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java (hunk 7 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java]: Since the legacy generators are public classes, they can be publicly deprecated, so in the last commit the `DEPRECATED` bit for them still relies on the annotation, which IMO is the authoritative "source of truth".

For the 10 other algorithms, which are accessible only via `RandomFactoryGenerator`, we can rely on the info in `RandomProperties`.

Comment (jaikiran) [reply src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java]: The deprecation process of algorithms is unclear - whether it should be tied with the `@Deprecation` of a class that provides that algorithm. We will have more clarity if/when we do deprecate these algorithms. The specification in its current form doesn't tie it to the `@Deprecation` annotation, which is a good thing. So your proposed change in this PR looks fine to me, since it will still allow us to move away from the `@Deprecated` annotation check if we wanted to, in a subsequent release/change.

**Review by jaikiran — APPROVED** [review]
Thank you Raffaello for fixing this as well as considering the review suggestions. The latest state of this PR in `880138d7` looks good to me. I've also reviewed the linked CSR and that too looks fine to me.

File: src/java.base/share/classes/java/util/random/package-info.java [hunk 10]
```diff
@@ -125,8 +125,8 @@
  *
  * <h2>Choosing a Random Number Generator Algorithm</h2>
  *
- * <p> There are three groups of random number generator algorithm provided
- * in Java: the Legacy group, the LXM group, and the Xoroshiro/Xoshiro group.
+ * <p> Random number generator algorithms are organized in groups,
+ * as described <a href="package-summary.html#algorithms">below</a>.
```
Comment (jddarcy) [inline src/java.base/share/classes/java/util/random/package-info.java:129]: FYI, it is possible to link to an anchor in a package-info file using syntax like

`@link java.util.random##algorithms`

**Review by jaikiran — APPROVED** [review]
